instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/HttpAttributesExtractor.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.api.instrumenter;
+
+import io.opentelemetry.api.common.AttributesBuilder;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+/**
+ * Extractor of <a
+ * href="https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/http.md">HTTP
+ * attributes</a>. Instrumentation of HTTP server or client frameworks should extend this class,
+ * defining {@link REQUEST} and {@link RESPONSE} with the actual request / response types of the
+ * instrumented library. If an attribute is not available in this library, it is appropriate to
+ * return {@code null} from the protected attribute methods, but implement as many as possible for
+ * best compliance with the OpenTelemetry specification.
+ */
+public abstract class HttpAttributesExtractor<REQUEST, RESPONSE>
+    extends AttributesExtractor<REQUEST, RESPONSE> {
+
+  @Override
+  protected final void onStart(AttributesBuilder attributes, REQUEST request) {
+    set(attributes, SemanticAttributes.HTTP_METHOD, method(request));
+    set(attributes, SemanticAttributes.HTTP_URL, url(request));
+    set(attributes, SemanticAttributes.HTTP_TARGET, target(request));
+    set(attributes, SemanticAttributes.HTTP_HOST, host(request));
+    set(attributes, SemanticAttributes.HTTP_ROUTE, route(request));
+    set(attributes, SemanticAttributes.HTTP_SCHEME, scheme(request));
+    set(attributes, SemanticAttributes.HTTP_USER_AGENT, userAgent(request));
```

Comment:
For this API, we just set what was provided indeed, I don't think we can model something like that at this layer since we don't know what can be inferred, or if we do then it needs to be more explicit in the spec I guess, that spec language is way too vague to be actually implementable I think.

Reply:
WDYM, vague? Spec gives several options for required attribute combinations and then warns agains duplication.

Reply:
Well these attributes are defined by the spec so it could say exactly which attributes should be filled. If it's the three sets of attributes and says attributes SHOULD NOT be filled from one set if another is filled, it would be clearer. I don't know either if this is a ordered list by priority. So vague ;)

Reply:
E.g, in that text server name is mentioned descriptively but it could just be said it should always be filled when available.